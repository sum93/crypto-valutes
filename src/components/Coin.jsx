import styled from 'styled-components'
import { useState } from 'react'

import Breakpoint from '../constants/Breakpoint'
import CoinDetails from './CoinDetails'
import Color from '../constants/Color'
import { currencyFormatter } from '../utils'
import { useCoins } from '../hooks/useCoins'

const CoinWrapper = styled.div`
display: grid;
grid-template-columns: min-content 4rem 1fr 1fr 5rem;
grid-template-rows: 1fr 1fr;
grid-template-areas:
  "logo symbol name current high"
  "logo symbol name current low"
  "details details details details details";
column-gap: 1rem;
width: 100%;
padding: 0.5rem 1.25rem;
border-radius: 1rem;
color: ${Color.DimGray};
background-color: ${Color.White};
cursor: pointer;

@media (max-width: ${Breakpoint.SM}) {
  grid-template-columns: 4rem 1fr 1fr;
  grid-template-rows: repeat(4, auto);
  grid-template-areas:
    "logo name high"
    "logo name current"
    "logo symbol current"
    "logo symbol low"
    "details details details";
  padding: 1.25rem;
}
`

const CoinLogo = styled.img`
grid-area: logo;
align-self: center;
height: 2.5rem;
width: 2.5rem;

@media (max-width: ${Breakpoint.SM}) {
  height: 4rem;
  width: 4rem;
}
`

const CoinSymbol = styled.span`
grid-area: symbol;
justify-self: end;
align-self: center;
font-size: 1.25rem;

@media (max-width: ${Breakpoint.SM}) {
  justify-self: start;
  align-self: start;
}
`

const CoinName = styled.span`
grid-area: name;
align-self: center;
font-size: 1.25rem;

@media (max-width: ${Breakpoint.SM}) {
  align-self: end;
}
`

const CoinCurrent = styled.span`
grid-area: current;
justify-self: end;
align-self: center;
font-size: 1.25rem;

@media (max-width: ${Breakpoint.SM}) {
  font-size: 1rem;
}
`

const CoinHigh = styled.span`
grid-area: high;
justify-self: start;
align-self: center;
font-size: 0.875rem;
color: ${Color.ShinyShamrock};

@media (max-width: ${Breakpoint.SM}) {
  justify-self: end;
  align-self: start;
}
`

const CoinLow = styled.span`
grid-area: low;
justify-self: start;
align-self: center;
font-size: 0.875rem;
color: ${Color.DarkSalmon};

@media (max-width: ${Breakpoint.SM}) {
  justify-self: end;
  align-self: end;
}
`

const Coin = ({ id }) => {
  const [isOpen, setIsOpen] = useState(false)
  const { coins: { data: { [id]: coin } } } = useCoins()

  const handleClick = () => {
    setIsOpen(prevIsOpen => !prevIsOpen)
  }

  if (!coin) {
    return null
  }

  return (
    <CoinWrapper onClick={handleClick}>
      <CoinLogo src={coin.image} />
      <CoinSymbol>{`[${coin.symbol}]`}</CoinSymbol>
      <CoinName>{coin.name}</CoinName>
      <CoinCurrent>{currencyFormatter.format(coin.current_price)}</CoinCurrent>
      <CoinHigh>{currencyFormatter.format(coin.high_24h)}</CoinHigh>
      <CoinLow>{currencyFormatter.format(coin.low_24h)}</CoinLow>
      {isOpen && <CoinDetails id={id} />}
    </CoinWrapper>
  )
}

export default Coin
