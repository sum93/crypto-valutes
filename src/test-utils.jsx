/* eslint-disable import/export */

import { render } from '@testing-library/react'
import { setupServer } from 'msw/node'

import { CoinsProvider } from './hooks/useCoins'
import { errorHandlers } from './mocks/errorHandlers'
import { handlers } from './mocks/handlers'

export const server = setupServer(...handlers)
export const useErrorHandlers = () => server.use(...errorHandlers)

const AllTheProviders = ({ children }) => {
  return (
    <CoinsProvider>
      {children}
    </CoinsProvider>
  )
}

const customRender = (ui, options) =>
  render(ui, { wrapper: AllTheProviders, ...options })

export * from '@testing-library/react'
export { customRender as render }
