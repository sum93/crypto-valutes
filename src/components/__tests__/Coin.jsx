import Coin from '../Coin'
import { render, screen, waitFor, fireEvent, waitForElementToBeRemoved } from '../../test-utils'

describe('<Coin />', () => {
  it('should render correctly', async () => {
    render(<Coin id="bitcoin" />)
    await waitFor(() => screen.findByTestId('coin-wrapper'))
    expect(screen.getByTestId('coin-wrapper')).toMatchSnapshot()
  })

  it('should display <CoinDetails /> when clicked on', async () => {
    render(<Coin id="bitcoin" />)
    await waitFor(() => screen.findByTestId('coin-wrapper'))
    expect(screen.getByTestId('coin-wrapper')).toBeInTheDocument()

    fireEvent.click(screen.getByTestId('coin-wrapper'))
    await waitForElementToBeRemoved(() => screen.queryByTestId('coin-details--loading'))
    expect(screen.getByTestId('coin-details')).toBeInTheDocument()
  })
})
