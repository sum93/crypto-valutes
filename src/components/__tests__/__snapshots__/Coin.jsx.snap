// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Coin /> should render correctly 1`] = `
.c0 {
  display: grid;
  grid-template-columns: min-content 4rem 1fr 1fr 5rem;
  grid-template-rows: 1fr 1fr;
  grid-template-areas: "logo symbol name current high" "logo symbol name current low" "details details details details details";
  -webkit-column-gap: 1rem;
  column-gap: 1rem;
  width: 100%;
  padding: 0.5rem 1.25rem;
  border-radius: 1rem;
  color: #62616b;
  background-color: #FCFCFC;
  cursor: pointer;
}

.c1 {
  grid-area: logo;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  height: 2.5rem;
  width: 2.5rem;
}

.c2 {
  grid-area: symbol;
  justify-self: end;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  font-size: 1.25rem;
}

.c3 {
  grid-area: name;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  font-size: 1.25rem;
}

.c4 {
  grid-area: current;
  justify-self: end;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  font-size: 1.25rem;
}

.c5 {
  grid-area: high;
  justify-self: start;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  font-size: 0.875rem;
  color: #71B48D;
}

.c6 {
  grid-area: low;
  justify-self: start;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  font-size: 0.875rem;
  color: #EEA081;
}

@media (max-width:600px) {
  .c0 {
    grid-template-columns: 4rem 1fr 1fr;
    grid-template-rows: repeat(4,auto);
    grid-template-areas: "logo name high" "logo name current" "logo symbol current" "logo symbol low" "details details details";
    padding: 1.25rem;
  }
}

@media (max-width:600px) {
  .c1 {
    height: 4rem;
    width: 4rem;
  }
}

@media (max-width:600px) {
  .c2 {
    justify-self: start;
    -webkit-align-self: start;
    -ms-flex-item-align: start;
    align-self: start;
  }
}

@media (max-width:600px) {
  .c3 {
    -webkit-align-self: end;
    -ms-flex-item-align: end;
    align-self: end;
  }
}

@media (max-width:600px) {
  .c4 {
    font-size: 1rem;
  }
}

@media (max-width:600px) {
  .c5 {
    justify-self: end;
    -webkit-align-self: start;
    -ms-flex-item-align: start;
    align-self: start;
  }
}

@media (max-width:600px) {
  .c6 {
    justify-self: end;
    -webkit-align-self: end;
    -ms-flex-item-align: end;
    align-self: end;
  }
}

<div
  class="c0"
  data-testid="coin-wrapper"
>
  <img
    class="c1"
    src="https://assets.coingecko.com/coins/images/1/large/bitcoin.png?1547033579"
  />
  <span
    class="c2"
  >
    [btc]
  </span>
  <span
    class="c3"
  >
    Bitcoin
  </span>
  <span
    class="c4"
  >
    €54,712.00
  </span>
  <span
    class="c5"
  >
    €54,796.00
  </span>
  <span
    class="c6"
  >
    €52,164.00
  </span>
</div>
`;
